docs/proposals/20231122-automate-aks-features.md

```diff
@@ -0,0 +1,428 @@
+reduces CAPZ's exposure to potential incompatibilities with certain ASO API versions.
+
+Regarding ClusterClass, this option functions the same as [Option 2] or [Option 3], where all ASO fields can
+be defined in a template. This option opens up an additional safeguard though, where webhooks could flag
+fields which should not be defined in a template. Similar webhook checks would be less practical in the other
+options where CAPZ would need to be aware of the set of disallowed fields for each ASO API version that a user
+could use.
+
+Similarly, CAPZ's webhooks are also better able to validate and default the ASO configuration and ensure
+fields like ManagedCluster's `spec.owner` that should not be modified by users are set correctly.
+
+#### Option 5: No change: CAPZ resource evolution proceeds the way it currently does
+
+This method describes not making any changes to how the CAPZ API is generally structured for AKS resources.
+CAPZ API types will continue to be curated manually without inheriting anything from the ASO API.
+
+Benefits of continuing on our current path include:
+- Familiarity with the existing pattern by users and contributors
+- Zero up-front cost to implement or transition to a new pattern
+- No requirement for users to have ASO knowledge
+- Greater freedom to change API implementations which we've recently leveraged to transition between the older
+  and newer Azure SDKs and to ASO.
+
+#### Option 6: Generate CAPZ code equivalent to what's added manually today
+
+For this option, a new code generation pipeline would be created to automatically scaffold the code that is
+currently manually written to expose additional AKS API fields to CAPZ.
+
+Once implemented, this method would drastically reduce the amount of developer effort to get started adding
+new AKS features. There would continue to be some amount of ongoing cost though to identify and handle nuances
+for each feature though, which may include issues outside of CAPZ's control like AKS API quirks.
+
+The main drawback of this approach from a developer perspective is the up-front effort to implement and
+ongoing cost to maintain the code generation itself. The existing AKS feature set exposed by CAPZ provides a
+decent foundation that can help identify regressions by testing the pipeline against features already
+implemented and tested. ASO also already implements a full code generation pipeline to transform Azure API
+specs into Kubernetes resource definitions, so some of that could possibly be reused by CAPZ.
+
+#### Option 7: CAPZ resource defines patches to ASO resource
+
+This method describes adding a new `spec.asoManagedClusterPatches` field to the existing
+AzureManagedControlPlane:
+
+```go
+type AzureManagedControlPlaneClassSpec struct {
+	...
+
+	// ASOManagedClusterPatches defines patches to be applied to the generated ASO ManagedCluster resource.
+	ASOManagedClusterPatches []string `json:"asoManagedClusterPatches,omitempty"`
```

Comment:
@dtzar and others: thoughts on gating this in the webhooks behind a new feature flag since it might be easy to misuse and/or break clusters?

Reply:
The plan in its current design would be to keep it in a permanent "break glass / best effort" category.  So, I'm torn because it's a bit of a user pain to have discover and remember to enable the user flag.  At the same time we need to communicate to users this is use at your own benefit or peril.  I also feel better in that if a user does something bad here and then removes it, *theoretically* the cluster should eventually go back to a healthy state.

If we can state it in docs and release notes without a feature flag, perhaps that's enough?